openassessment/assessment/test/test_staff.py

```diff
@@ -522,3 +523,238 @@ def _create_student_and_submission(student, answer, date=None, problem_steps=Non
+        that validation, beware.
+        """
+        # pylint: disable=unexpected-keyword-arg
+        return cls.create_workflow(
+            course_id=course_id or cls.course_id,
+            item_id=item_id or cls.item_id,
+            grading_completed_at=None,
+            cancelled_at=None,
+            scorer_id=scorer_id,
+            grading_started_at=grading_started_at,
+        )
+
+    @classmethod
+    def _create_graded(cls, course_id=None, item_id=None, scorer_id=''):
+        """
+        Create a completed, graded workflow (no assessment is created)
+        """
+        # pylint: disable=unexpected-keyword-arg
+        return cls.create_workflow(
+            course_id=course_id or cls.course_id,
+            item_id=item_id or cls.item_id,
+            scorer_id=scorer_id,
+            cancelled_at=None,
+            grading_completed_at=datetime.now()
+        )
+
+    def test_cancelled(self):
+        workflow = self.create_workflow()
+        self.assertFalse(workflow.is_cancelled)
+        workflow.cancelled_at = datetime.now()
+        self.assertTrue(workflow.is_cancelled)
+
+    @unpack
+    @data(
+        (0, 5, 7),
+        (3, 0, 4),
+        (5, 2, 0),
+        (0, 0, 0),
+    )
+    def test_get_workflow_statistics(self, expected_graded, expected_ungraded, expected_in_progress):
+        for _ in range(expected_graded):
+            self._create_graded()
+        for _ in range(expected_ungraded):
+            self._create_ungraded()
+        for _ in range(expected_in_progress):
+            self._create_in_progress()
+        stats = self.model.get_workflow_statistics(self.course_id, self.item_id)
+        self.assertDictEqual(
+            {
+                'graded': expected_graded,
+                'ungraded': expected_ungraded,
+                'in-progress': expected_in_progress,
+            },
+            stats
+        )
+
+    def _get_and_assert_workflow(self, expected_workflow):
+        """
+        Call get_submission_for_review for course_id, item_id, and scorer_1_id
+        Verify the identifying uuid returned is the expected workflow, and
+        check that scorer_id and grading_started_at were updated
+        """
+
+        submission_uuid = self.model.get_submission_for_review(self.course_id, self.item_id, self.scorer_1_id)
+        selected_workflow = self.get_workflow_by_identifying_uuid(submission_uuid)
+
+        # Check that the workflow is the workflow we expect, and that scorer_id was updated, and that
+        # grading_started_at was set no more than a second ago.
+        self.assertEqual(expected_workflow.id, selected_workflow.id)
+        self.assertEqual(self.scorer_1_id, selected_workflow.scorer_id)
+
+        now_with_tz = datetime.now(tz=selected_workflow.grading_started_at.tzinfo)
+        timediff = now_with_tz - selected_workflow.grading_started_at
+        self.assertLess(timediff, timedelta(seconds=1))
```

Comment:
Is this to test performance or could we mock the time to better test this?

Reply:
Nope, not meant to be a performance check, I just wanted to avoid having to mock time xD

Reply:
We use freezegun in lots of our repos to freeze time for testing: https://pypi.org/project/freezegun/0.1.11/